File: chap4_imple/codetree_middle_bingbing.py
'''
2023.11.02
https://www.codetree.ai/missions/5/problems/snail-start-from-center?&utm_source=clipboard&utm_medium=text
직사각형 가운데에서 시작해서 빙빙 돌며 숫자 채우기
'''
n = int(input())

a = [[1 for _ in range(n)] for _ in range(n)]
num = 1
x, y = n//2, n//2
r_cnt, c_cnt = 1,1
cur_d = 0
dirs = ((0,1),(-1,0),(0,-1),(1,0))     # 오,위,왼,아
while True:
    if num == n*n: break
    if cur_d % 2 == 0:
        for _ in range(c_cnt):
            num += 1
            x, y = x + dirs[cur_d][0], y + dirs[cur_d][1]
            a[x][y] = num
            if num == n*n: break
        c_cnt += 1
    else:
        for _ in range(r_cnt):
            num += 1
            x, y = x + dirs[cur_d][0], y + dirs[cur_d][1]
            a[x][y] = num
            if num == n*n: break
        r_cnt += 1
    
    cur_d = (cur_d+1)%4

for i in range(n):
    for j in range(n):
        print(a[i][j], end=' ')
    print()
    
'''
해설
'''

# 변수 선언 및 입력:
n = int(input())
grid = [
    [0 for _ in range(n)]
    for _ in range(n)
]

# 시작 위치와 방향,
# 해당 방향으로 이동할 횟수를 설정
curr_x, curr_y = n // 2, n // 2
move_dir, move_num = 0, 1
def in_range(x,y):
    return 0 <= x and x < n and 0 <= y and y < n

# 한 칸 움직이면서 청소를 진행
def move():
    global curr_x, curr_y   # 현재 위치
    
    # 문제에서 원하는 진행 순서대로
    # 오른쪽, 위, 왼쪽, 아래 방향이 되도록 정의하기
    dxs, dys = [0,-1,0,1], [1,0,-1,0]
    curr_x, curr_y = curr_x+dxs[move_dir], curr_y+dys[move_dir]

def end():
    return not in_range(curr_x, curr_y)

# 시작 위치와 방향,
# 해당 방향으로 이동할 횟수를 설정
cnt = 1

while not end():
    # move_num만큼 이동 
    for _ in range(move_num):
        grid[curr_x][curr_y] = cnt
        cnt += 1
        move()
        
        # 이동하는 도중 격자를 벗어나면 종료
        if end(): break
    
    # 방향을 바꿈
    move_dir = (move_dir+1)%4
    
    # 만약 현재 방향이 왼쪽/오른쪽이 된 경우
    # 특정 방향으로 움직여야 할 횟수를 1 증가
    if move_dir == 0 or move_dir == 2:
        move_num += 1

# 출력:
for i in range(n):
    for j in range(n):
        print(grid[i][j], end=' ')
    print()

File: chap4_imple/codetree_middle_bingbing.cpp
/*
 * 2023.11.12
 * 마지막에 1,2,3,4,4가 된다는 것을 놓침
 * 내 풀이
*/
#include <iostream>
#define MAX_N 100

using namespace std;

int n, x, y;                                    // n: 격자 크기, (x, y): 초기 좌표
int c_cnt = 1, r_cnt = 1;                       // c_cnt: 오/왼 방향으로 채울 개수, r_cnt: 위/아래 방향으로 채울 개수
int dirs[4][2] = {{0,1},{-1,0},{0,-1},{1,0}};   // 우, 상, 좌, 하 방향
int cur_d;                                      // 현재 방향 인덱스
int arr[MAX_N][MAX_N];                          // 채울 격자
int val = 1;                                    // 격자에 채워지는 값

void write_num(int cnt){    // cnt 개수 만큼 값을 채움
    for(int i=0; i<cnt; i++){
        x = x+dirs[cur_d][0], y = y+dirs[cur_d][1];     // cur_d 방향으로 cnt만큼 값을 채움
        if(val+1 > n*n) return;     // 만약 숫자를 모두 채웠다면 리턴
        arr[x][y] = ++Val;
    }
}

int main(){
    cin >> n;
    x = n/2, y = n/2;
    arr[x][y] = val;    // 가운데에 숫자 1 채우기
    while(true){
        if(cur_d % 2 == 0){     // 우/좌 방향이라면
            write_num(c_cnt);   // c_cnt만큼 숫자 채우기
            c_cnt++;            // 채워야 할 개수 증가
        }
        else{                   // 상/하 방향이라면
            write_num(r_cnt);   // r_cnt만큼 숫자 채우기
            r_cnt++;            // 채워야 할 개수 증가 
        }
        cur_d = (cur_d+1) % 4;
    }

    // 출력
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cout << arr[i][j] << ' ';
        }
        cout<<'\n';
    }
    return 0;
}